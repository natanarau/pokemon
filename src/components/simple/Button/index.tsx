import { BoxProps } from 'components/simple/Box';
import { LoadingIcon } from 'components/simple/LoadingIcon';

import * as S from './styles';

export type ButtonProps = BoxProps & {
  variant?: 'primary' | 'secondary' | 'ghost' | 'white' | 'green' | 'dark' | 'blue' | 'purple';
  border?: 'round' | 'square';
  children: React.ReactNode;
  as?: 'a' | 'button';
  onClick?: () => void;
  disabled?: boolean | any;
  falseDisabled?: boolean | any;
  loading?: boolean;
};

const Button = ({
  variant = 'primary',
  border = 'round',
  loading,
  children,
  ...rest
}: ButtonProps) => (
  <S.Wrapper data-testid="Button" {...rest} border={border} variant={variant}>
    {loading ? <LoadingIcon /> : children}
  </S.Wrapper>
);

export default Button;
