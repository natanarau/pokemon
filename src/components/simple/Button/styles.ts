import styled, { css } from 'styled-components';
import { switchProp, ifProp, theme } from 'styled-tools';

import { Box } from 'components/simple/Box';

export const Wrapper = styled(Box)<any>`
  cursor: pointer;
  text-align: center;
  font-size: ${theme('fontSizes.md')};
  display: flex;
  align-items: center;
  justify-content: ${({ justifyContent }) => justifyContent || 'center'};
  border: none;

  ${switchProp('variant', {
    primary: css`
      color: ${theme('colors.realWhite')};
      background-color: ${theme('colors.petroleum.100')};
    `,
    secondary: css`
      color: ${theme('colors.petroleum.100')};
      background-color: transparent;
      border: 1px solid ${theme('colors.petroleum.100')};
    `,
    ghost: css`
      border: 0;
      color: ${theme('colors.petroleum.100')};
      background-color: transparent;
    `,
    white: css`
      border: 1px solid ${theme('colors.petroleum.100')};
      color: ${theme('colors.petroleum.100')};
      background-color: #fff;
    `,
    dark: css`
      border: 0;
      color: #47c8cd;
      background-color: #000;
    `,
    green: css`
      color: #3A7426;
      background-color: #A8F083;
      border: 1px solid #82c56b;
      font-weight: 800;
      font-size: .8rem;
    `,
    blue: css`
      color: #07286A;
      background-color: #98A8FF;
      border: 1px solid #7e90b5;
      font-weight: 800;
      font-size: .8rem;
    `,
    purple: css`
      color: #463338;
      background-color: #ACABCB;
      border: 1px solid #877db5;
      font-weight: 800;
      font-size: .8rem;
    `
  })}
  ${switchProp('border', {
    round: css`
      border-radius: .4rem;
    `,
    square: css`
      border-radius: 0;
    `,
  })}
  transition: all .4s cubic-bezier(0.075, 0.82, 0.165, 1);
  :hover {
    opacity: 0.9;
  }
  ${ifProp(
    'disabled',
    css`
      background-color: #606060;
      opacity: 0.7;
      cursor: not-allowed;
      border-color: #606060;
      :hover {
        filter: brightness(1);
      }
    `
  )}
  ${ifProp(
    'falseDisabled',
    css`
      background-color: #606060;
      opacity: 0.7;
      border-color: #606060;
      :hover {
        filter: brightness(1);
      }
    `
  )}
`;
