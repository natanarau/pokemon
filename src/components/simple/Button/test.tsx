import { render, waitFor, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import Button from '.';

afterEach(() => {
  cleanup();
  jest.resetModules();
});

describe('<Button />', () => {
  test('should render icon correctly', async () => {
    const { getByTestId } = render(<Button as='button' variant='primary'>Button</Button>)
    await waitFor(() => expect(getByTestId('Button')).toBeTruthy())
  })
  test('should click', async () => {
    const onClick = jest.fn()
    const { getByTestId } = render(<Button as='button' onClick={onClick} variant='primary'>Button</Button>)
    const button = await waitFor(() => getByTestId('Button'))
    userEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(1)
  })
  test('should match snapshot', () => {
    const { container } = render(
      <Button as='button' variant='primary'>Button</Button>
    )
    expect(container).toMatchSnapshot();
  });
})
