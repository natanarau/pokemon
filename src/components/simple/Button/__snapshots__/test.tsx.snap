// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
.c0 {
  border: round;
}

.c1 {
  cursor: pointer;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: none;
  border-radius: 5rem;
  -webkit-transition: all .4s cubic-bezier(0.075,0.82,0.165,1);
  transition: all .4s cubic-bezier(0.075,0.82,0.165,1);
}

.c1:hover {
  opacity: 0.9;
}

<div>
  <button
    class="c0 c1"
    data-testid="Button"
  >
    Button
  </button>
</div>
`;
